datalad/distribution/clone.py

```diff
@@ -0,0 +1,249 @@
+            # and derive the path from the source and continue
+            path = _get_installationpath_from_url(source)
+            # since this is a relative `path`, resolve it:
+            path = resolve_path(path, dataset)
+            lgr.debug("Determined clone target path from source")
+        lgr.debug("Resolved clone target path to: '%s'", path)
+
+        # there is no other way -- my intoxicated brain tells me
+        assert(path is not None)
+
+        destination_dataset = Dataset(path)
+        dest_path = path
+
+        status_kwargs = dict(
+            action='clone', ds=destination_dataset, logger=lgr,
+            refds=refds_path)
+
+        # important test! based on this `rmtree` will happen below after failed clone
+        if exists(dest_path) and listdir(dest_path):
+            if destination_dataset.is_installed():
+                # check if dest was cloned from the given source before
+                # this is where we would have installed this from
+                guessed_sources = _get_flexible_source_candidates(
+                    source, dest_path)
+                # this is where it was actually installed from
+                track_name, track_url = _get_tracking_source(destination_dataset)
+                if track_url in guessed_sources or \
+                        get_local_file_url(track_url) in guessed_sources:
+                    yield get_status_dict(
+                        status='notneeded',
+                        message=("dataset %s was already cloned from '%s'",
+                                 destination_dataset,
+                                 source),
+                        **status_kwargs)
+                    return
+            # anything else is an error
+            yield get_status_dict(
+                status='error',
+                message='target path already exists and not empty, refuse to clone into target path',
+                **status_kwargs)
+            return
+
+        if dataset is not None and relpath(path, start=dataset.path).startswith(pardir):
+            yield get_status_dict(
+                status='error',
+                message=("clone target path '%s' not in specified target dataset '%s'",
+                         path, dataset),
+                **status_kwargs)
+            return
+
+        # generate candidate URLs from source argument to overcome a few corner cases
+        # and hopefully be more robust than git clone
+        candidate_sources = _get_flexible_source_candidates(source)
+        for source_ in candidate_sources:
+            try:
+                lgr.info("Attempting to clone dataset from '%s' to '%s'",
+                         source_, dest_path)
+                # TODO this is a shame. we should be able to instruct GitRepo
+                # how to behave wrt logging, and not stick a pillow in its face
+                with swallow_logs():
```

Comment:
FTR: The swallowing was in-place before (it just appears as an addition due to other RF). I merely added the comment.